### DISCUSSION TITLE
Reading a tag value along with the PLC time

### DISCUSSION THREAD

**Discussion by AlexRegev** [ottowayi/pycomm3#175] [Q&A] [answered]
Hi,

Is there a way to read a tag value along with the PLC time to which it corresponds? 
I am trying to monitor a value in the PLC and I need to be able to have a very accurate timestamp for that value (using the PC clock is not accurate enough).

An example for what I am looking for would be the way OPC-UA provides the server source time when reading a node, is there something parallel to that in CIP?
Perhaps there is some method that can be used to create a similar functionality?

Any guidance would be very much appreciated,

Alex

**Comment by ottowayi** [comment, marked as answer]
It's not part of the CIP protocol, but you can easily add the functionality using the `get_plc_time` method.  There are many ways to do it, but here is a short example that adds the timestamp by subclass the `LogixDriver` and adds a `read_with_timestamp` method:

```
class MyLogixDriver(LogixDriver):

    def read_with_timestamp(self, *tags):
        plc_time = self.get_plc_time().value["string"]
        results = self.read(*tags)

        return [(*result, plc_time) for result in results]
```

```
with MyLogixDriver("192.168.1.236") as plc:
    tags = plc.read_with_timestamp("read_dint1", "read_sint1")

print(tags)
```

Output
```
[('read_dint1', 0, 'DINT', None, 'Wednesday, October 20, 2021 09:07:29PM'), ('read_sint1', 0, 'SINT', None, 'Wednesday, October 20, 2021 09:07:29PM')]
```

**Reply by AlexRegev** [reply]
Hi Ian,

Thank you so much for the very prompt response.
I am wondering if there is a way to "atomically" read the plc time along with the data, I am worried that since the time and the data are read in two different function calls then probably there will be a few milliseconds of difference between them.

Your answer lead me to a thought, I was wondering if you think this could work:
1. The PLC will have a tag in which the PLC time will be stored and updated.
2. the "read_with_timestamp" function will read the requested data along with the tag containing the PLC time for each packet exchanged with the PLC (packet size dependent on the PLC type as described in your manual)
3. Since the PLC would be handling the data and time read in a single transaction this would result in high accuracy of reading the PLC time along with the data.

Do you think this could work?

Thanks again for your help and guidance.

Alex

**Reply by ottowayi** [reply]
yeah, that could work.. but you will also have latency in your PLC moving the time into a tag.  That tag will be updated at the scan time of your program, so it could be out of sync and even slower than the couple ms delay between the two network calls.  It really depends on what resolution you need.
